src/EventConsumerService.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) Sven Rosvall (sven@rosvall.ie)
+// This file is part of VLCB-Arduino project on https://github.com/SvenRosvall/VLCB-Arduino
+// Licensed under the Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License.
+// The full licence can be found at: http://creativecommons.org/licenses/by-nc-sa/4.0/
+
+#include <Streaming.h>
+
+#include "EventConsumerService.h"
+#include <Controller.h>
+#include <vlcbdefs.hpp>
+
+namespace VLCB {
+
+void EventConsumerService::setController(Controller *cntrl) 
+{
+  this->controller = cntrl;
+  this->module_config = cntrl->module_config;
+}
+
+//
+/// register the user handler for learned events
+//
+void EventConsumerService::setEventHandler(void (*fptr)(byte index, CANFrame *msg)) 
+{
+  eventhandler = fptr;
+}
+
+// overloaded form which receives the opcode on/off state and the first event variable
+void EventConsumerService::setEventHandler(void (*fptr)(byte index, CANFrame *msg, bool ison, byte evval)) 
+{
+  eventhandlerex = fptr;
+}
+
+//
+/// for accessory event messages, lookup the event in the event table and call the user's registered event handler function
+//
+void EventConsumerService::processAccessoryEvent(CANFrame *msg, unsigned int nn, unsigned int en, bool is_on_event) 
+{
+  // try to find a matching stored event -- match on nn, en
+  byte index = module_config->findExistingEvent(nn, en);
+
+  // call any registered event handler
+
+  if (index < module_config->EE_MAX_EVENTS) {
+    if (eventhandler != NULL) {
+      (void)(*eventhandler)(index, msg);
+    } else if (eventhandlerex != NULL) {
+      byte evVal = (module_config->EE_NUM_EVS > 0) ? module_config->getEventEVval(index, 1) : 0;
+      (void)(*eventhandlerex)(index, msg, is_on_event, evVal);
+    }
+  }
+}
+
+Processed EventConsumerService::handleMessage(unsigned int opc, CANFrame *msg) 
+{
+  unsigned int nn = (msg->data[1] << 8) + msg->data[2];
+  unsigned int en = (msg->data[3] << 8) + msg->data[4];
+  // DEBUG_SERIAL << ">CbusSvc handling message op=" << _HEX(opc) << " nn=" << nn << " en" << en << endl;
+
+  switch (opc) 
+  {
+    case OPC_ACON:
+    case OPC_ACON1:
+    case OPC_ACON2:
+    case OPC_ACON3:
+
+    case OPC_ACOF:
+    case OPC_ACOF1:
+    case OPC_ACOF2:
+    case OPC_ACOF3:
```

Comment:
There is a discussion if these messages with data should be used with the event handler that doesn't handle that data. I am happy to leave as is for now and fix the ones with data when we know what to do with them.